File: mcp-code-intelligence-kotlin/src/test/kotlin/com/codeintel/memory/search/AgentScopeFilterTest.kt
/** Unit tests for AgentScopeFilter — tag-based KB isolation. */
package com.codeintel.memory.search

import com.codeintel.memory.models.KnowledgeEntry
import com.codeintel.memory.models.KnowledgeSearchResult
import java.sql.DriverManager
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class AgentScopeFilterTest {

    private fun createDb(): java.sql.Connection {
        val conn = DriverManager.getConnection("jdbc:sqlite::memory:")
        conn.createStatement().execute("""
            CREATE TABLE agent_scope_config (
              id INTEGER PRIMARY KEY AUTOINCREMENT,
              agent_role TEXT NOT NULL UNIQUE,
              tag_set TEXT NOT NULL DEFAULT '[]',
              updated_at TEXT NOT NULL DEFAULT (datetime('now'))
            )
        """.trimIndent())
        conn.createStatement().execute("""
            INSERT INTO agent_scope_config (agent_role, tag_set) VALUES
              ('QA', '["testing","qa","test-plan","bug"]'),
              ('DEV', '["code","api","architecture","design"]'),
              ('BA', '["requirement","business","process"]')
        """.trimIndent())
        return conn
    }

    private fun makeResult(tags: String, id: Long = 1): KnowledgeSearchResult {
        return KnowledgeSearchResult(
            entry = KnowledgeEntry(id = id, content = "test", summary = "test", type = "CONTEXT", tags = tags),
            score = 1.0
        )
    }

    @Test
    fun `getScope returns correct tags for known role`() {
        val conn = createDb()
        val filter = AgentScopeFilter(conn)
        val scope = filter.getScope("QA")
        assertNotNull(scope)
        assertEquals("QA", scope.role)
        assertTrue(scope.tags.contains("testing"))
        assertTrue(scope.tags.contains("bug"))
    }

    @Test
    fun `getScope returns null for unknown role`() {
        val conn = createDb()
        val filter = AgentScopeFilter(conn)
        assertNull(filter.getScope("UNKNOWN"))
    }

    @Test
    fun `filter keeps entries matching agent tags`() {
        val conn = createDb()
        val filter = AgentScopeFilter(conn)
        val results = listOf(
            makeResult("testing,qa", 1),
            makeResult("code,api", 2),
            makeResult("requirement", 3),
        )
        val filtered = filter.filter(results, "QA")
        assertEquals(1, filtered.size)
        assertEquals(1L, filtered[0].entry.id)
    }

    @Test
    fun `filter keeps untagged entries for all roles`() {
        val conn = createDb()
        val filter = AgentScopeFilter(conn)
        val results = listOf(
            makeResult("", 1),  // untagged
            makeResult("code", 2),
        )
        val filtered = filter.filter(results, "QA")
        assertEquals(1, filtered.size) // only untagged passes
        assertEquals(1L, filtered[0].entry.id)
    }

    @Test
    fun `filter returns all results for unknown role`() {
        val conn = createDb()
        val filter = AgentScopeFilter(conn)
        val results = listOf(makeResult("testing", 1), makeResult("code", 2))
        val filtered = filter.filter(results, "UNKNOWN")
        assertEquals(2, filtered.size)
    }

    @Test
    fun `updateScope persists and updates cache`() {
        val conn = createDb()
        val filter = AgentScopeFilter(conn)
        filter.updateScope("QA", listOf("testing", "qa", "e2e", "performance"))
        val scope = filter.getScope("QA")
        assertNotNull(scope)
        assertEquals(4, scope.tags.size)
        assertTrue(scope.tags.contains("e2e"))
    }

    @Test
    fun `case insensitive role lookup`() {
        val conn = createDb()
        val filter = AgentScopeFilter(conn)
        assertNotNull(filter.getScope("qa"))
        assertNotNull(filter.getScope("Qa"))
        assertNotNull(filter.getScope("QA"))
    }

    @Test
    fun `filter handles mixed case tags`() {
        val conn = createDb()
        val filter = AgentScopeFilter(conn)
        val results = listOf(makeResult("Testing,QA", 1))
        val filtered = filter.filter(results, "QA")
        assertEquals(1, filtered.size) // "testing" matches after lowercase
    }
}

File: backend/mcp-code-intelligence-nodejs/src/memory/tests/agent-scope-filter.test.ts
/**
 * Unit tests for AgentScopeFilter — tag-based KB isolation.
 */

import { describe, it, beforeEach } from 'node:test';
import assert from 'node:assert/strict';
import Database from 'better-sqlite3';
import { AgentScopeFilter } from '../v2/agent-scope-filter.js';
import { SearchResult, KnowledgeEntry } from '../models.js';

function makeEntry(id: number, tags: string): KnowledgeEntry {
  return {
    id, content: 'test', summary: 'test', type: 'CONTEXT', tier: 'WORKING',
    source: null, source_ref: null, tags, confidence: 1, access_count: 0,
    created_at: '', updated_at: '', last_accessed_at: null, expires_at: null,
  };
}

function makeResult(id: number, tags: string): SearchResult {
  return { entry: makeEntry(id, tags), score: 1.0, matchType: 'hybrid' };
}

describe('AgentScopeFilter', () => {
  let db: Database.Database;
  let filter: AgentScopeFilter;

  beforeEach(() => {
    db = new Database(':memory:');
    db.exec(`
      CREATE TABLE agent_scope_config (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        agent_role TEXT NOT NULL UNIQUE,
        tag_set TEXT NOT NULL DEFAULT '[]',
        updated_at TEXT NOT NULL DEFAULT (datetime('now'))
      );
      INSERT INTO agent_scope_config (agent_role, tag_set) VALUES
        ('QA', '["testing","qa","test-plan","test-case","bug"]'),
        ('DEV', '["code","api","architecture","implementation","design"]'),
        ('BA', '["requirement","business","stakeholder","process"]');
    `);
    filter = new AgentScopeFilter(db);
  });

  describe('getScope', () => {
    it('returns scope for known role', () => {
      const scope = filter.getScope('QA');
      assert(scope !== null);
      assert.equal(scope!.role, 'QA');
      assert(scope!.tags.includes('testing'));
    });

    it('returns null for unknown role', () => {
      assert.equal(filter.getScope('UNKNOWN'), null);
    });

    it('is case-insensitive', () => {
      const scope = filter.getScope('qa');
      assert(scope !== null);
      assert.equal(scope!.role, 'QA');
    });
  });

  describe('filter', () => {
    it('filters results to matching tags only', () => {
      const results = [
        makeResult(1, 'testing, qa'),
        makeResult(2, 'code, api'),
        makeResult(3, 'testing'),
      ];
      const filtered = filter.filter(results, 'QA');
      assert.equal(filtered.length, 2);
      assert.deepEqual(filtered.map(r => r.entry.id), [1, 3]);
    });

    it('keeps untagged entries visible to all (BR-F4-02)', () => {
      const results = [
        makeResult(1, 'testing'),
        makeResult(2, ''),  // untagged
        makeResult(3, 'code'),
      ];
      const filtered = filter.filter(results, 'QA');
      assert.equal(filtered.length, 2);
      assert.deepEqual(filtered.map(r => r.entry.id), [1, 2]);
    });

    it('returns all results for unknown role (no filtering)', () => {
      const results = [makeResult(1, 'testing'), makeResult(2, 'code')];
      const filtered = filter.filter(results, 'UNKNOWN');
      assert.equal(filtered.length, 2);
    });

    it('handles entries with multiple matching tags', () => {
      const results = [makeResult(1, 'testing, code, qa')];
      const filtered = filter.filter(results, 'QA');
      assert.equal(filtered.length, 1);
    });
  });

  describe('updateScope', () => {
    it('updates existing scope', () => {
      filter.updateScope('QA', ['testing', 'qa', 'e2e']);
      const scope = filter.getScope('QA');
      assert(scope!.tags.includes('e2e'));
      assert(!scope!.tags.includes('bug'));
    });

    it('creates new scope for unknown role', () => {
      filter.updateScope('PM', ['planning', 'roadmap']);
      const scope = filter.getScope('PM');
      assert(scope !== null);
      assert.deepEqual(scope!.tags, ['planning', 'roadmap']);
    });
  });
});
